pkg/exec/devfile.go

```diff
@@ -222,6 +224,61 @@ func ExecuteDevfileDebugActionWithoutRestart(client ExecClient, exec common.Exec
 	return nil
 }
 
+// ExecuteCompositeDevfileAction executes a given composite command in a devfile
+// The composite command may reference exec commands, composite commands, or both
+func ExecuteCompositeDevfileAction(client ExecClient, composite common.Composite, commandsMap map[string]common.DevfileCommand, compInfo adaptersCommon.ComponentInfo, show bool, machineEventLogger machineoutput.MachineEventLoggingClient) (err error) {
+	if composite.Parallel {
+		// Loop over each command and execute it in parallel
+		commandExecs := util.NewConcurrentTasks(len(composite.Commands))
+		for _, command := range composite.Commands {
+			cmd := command // needed to prevent the lambda from capturing the value
+			if devfileCommand, ok := commandsMap[strings.ToLower(cmd)]; ok {
+				commandExecs.Add(util.ConcurrentTask{ToRun: func(errChannel chan error) {
+					err := execCommandFromComposite(client, devfileCommand, commandsMap, compInfo, show, machineEventLogger)
+					if err != nil {
+						errChannel <- err
+					}
+				}})
+			} else {
+				return fmt.Errorf("command %q not found in devfile", command)
+			}
+		}
+
+		err := commandExecs.Run()
+		if err != nil {
+			return errors.Wrap(err, "command exec failed")
```

Comment:
Can we be more explicit here like, parallel execution of comp command `<name>` failed

Reply:
Good point